### PR TITLE
Split filter_note_type function and update search method

### PR DESCRIPTION
**Pull request by ehanson8**
#### What does this PR do?
Updates related to a new round of find & replace changes. The filter_note_type function has been split out into 3 separate functions with accompanying tests. The search method has been updated with an advanced query for greater flexibility and efficiency. The main function in cli.py will be adapted for a find & replace function in models.py in a subsequent PR

#### Includes new or updated dependencies?
NO

### CODE REVIEW

File: asaps/models.py [hunk 1]
```diff
@@ -52,10 +53,13 @@ def get_all_records(self, endpoint):
         ids = self.client.get(f'{endpoint}?all_ids=true').json()
         return ids
 
-    def search(self, string, repo_id, rec_type):
+    def search(self, string, repo_id, rec_type, note_type='keyword'):
         """Search for a string across a particular record type."""
-        endpoint = (f'repositories/{repo_id}/search?q="{string}'
-                    f'"&page_size=100&type[]={rec_type}')
+        endpoint = f'repositories/{repo_id}/search?'
```
Comment (gravesm) [inline asaps/models.py:58]: This is pretty difficult to read and seems likely to end up with accidental typos or improperly quoted characters. The asnake client isn't well documented, but if you look at the source you can see that they pass the `params` keyword argument in `get_paged()` on to the underlying requests call. I think what I'd do here is to pass all the query params in through a dictionary and for the `aq` param, start with it as a dictionary and convert it to a JSON string. This way you can just let python and requests deal with all the quoting. So something like:

```python
query = {'query': {'field': note_type, 'value': string, 'jsonmodel_type': 'field_query'}}
params = {'ag': json.dumps(query), 'page_size': 100, 'type[]': rec_type}
return self.client.get_paged(endpoint, params=params)
```

File: asaps/models.py [hunk 2]
```diff
@@ -121,40 +126,30 @@ def create_csv(csv_data, file_name):
     return full_file_name
 
 
-def filter_note_type(client, csv_data, rec_obj, note_type, operation,
-                     old_string='', new_string=''):
-    """Filter notes by type for exporting or editing.
+def filter_note_type(rec_obj, note_type):
```
Comment (gravesm) [inline asaps/models.py:129]: This suggestion, in this case, is purely stylistic, but a nice one-line pythonic solution for this is to use a generator expression:

```python
def filter_note(rec_obj, note_type):
    return (n for n in rec_obj['notes'] if n.get('type') == note_type)
```

File: asaps/models.py [hunk 3]
```diff
@@ -121,40 +126,30 @@ def create_csv(csv_data, file_name):
...
+    """Filter notes by type."""
+    for note_gen in [n for n in rec_obj['notes']
+                     if n.get('type') == note_type]:
+        yield note_gen
 
-    Triggers post of updated record if changes are made.
+
+def retrieve_subnotes(note):
```
Comment (gravesm) [inline asaps/models.py]: Unless there's more planned for this function, I'm not sure it's necessary. The difference in code would be:

```python
for subnote in retrieve_subnotes(note):
    # do things
```
vs
```python
for subnote in note.get('subnotes', []):
    # do things
```
The latter is only one more character to type and doesn't require someone reading the code to go lookup a function.
```

**Review by gravesm — CHANGES_REQUESTED** [review]
The `search()` method I think should be adjusted. The other two suggestions are optional.

File: asaps/models.py (hunk 1 above)
Comment (ehanson8) [reply asaps/models.py:58]: Great suggestion, this is much cleaner

**Review by gravesm — APPROVED** [review]

**Pull request merged**